Sources/NIO/EventLoopFuture.swift

```diff
@@ -643,14 +646,21 @@ extension EventLoopFuture {
     public func flatMapResult<NewValue, SomeError: Error>(file: StaticString = #file,
                                                           line: UInt = #line,
                                                           _ body: @escaping (Value) -> Result<NewValue, SomeError>) -> EventLoopFuture<NewValue> {
-        return self.flatMap(file: file, line: line) { value in
-            switch body(value) {
+        let next = EventLoopPromise<NewValue>(eventLoop: eventLoop, file: file, line: line)
+        self._whenComplete {
+            switch self._value! {
             case .success(let value):
-                return self.eventLoop.makeSucceededFuture(value, file: file, line: line)
-            case .failure(let error):
-                return self.eventLoop.makeFailedFuture(error, file: file, line: line)
+                switch body(value) {
+                case .success(let newValue):
+                    return next._setValue(value: .success(newValue))
+                case .failure(let error):
+                    return next._setValue(value: .failure(error))
+                }
```

Comment:
The `switch` here is unnecessary: this can just be `return next._setValue(value: body(value))`

Reply:
This btw explains why we also had a switch in the code we replace:

```swift
        return self.flatMap(file: file, line: line) { value in
            switch body(value) {
            case .success(let value):
                return self.eventLoop.makeSucceededFuture(value, file: file, line: line)
            case .failure(let error):
                return self.eventLoop.makeFailedFuture(error, file: file, line: line)
            }
        }
```

Reply:
Christ what a pain. Fair enough.